state/watcher.go

```diff
@@ -496,6 +497,185 @@ func watchApplicationRelations(backend modelBackend, applicationName string) Str
 	return newRelationLifeSuspendedWatcher(backend, members, filter, nil)
 }
 
+// WatchModelMachineStartTimes watches the non-container machines in the model
+// for changes to the Life or AgentStartTime fields and reports them as a batch
+// after the specified quiesceInterval time has passed without seeing any new
+// change events.
+func (st *State) WatchModelMachineStartTimes(quiesceInterval time.Duration) StringsWatcher {
+	return newModelMachineStartTimeWatcher(st, st.clock(), quiesceInterval)
+}
+
+type modelMachineStartTimeFieldDoc struct {
+	Id             string    `bson:"_id"`
+	Life           Life      `bson:"life"`
+	AgentStartedAt time.Time `bson:"agent-started-at"`
+}
+
+type modelMachineStartTimeWatcher struct {
+	commonWatcher
+	outCh chan []string
+
+	clk             clock.Clock
+	quiesceInterval time.Duration
+	seenDocs        map[string]modelMachineStartTimeFieldDoc
+}
+
+func newModelMachineStartTimeWatcher(backend modelBackend, clk clock.Clock, quiesceInterval time.Duration) StringsWatcher {
+	w := &modelMachineStartTimeWatcher{
+		commonWatcher:   newCommonWatcher(backend),
+		outCh:           make(chan []string),
+		clk:             clk,
+		quiesceInterval: quiesceInterval,
+		seenDocs:        make(map[string]modelMachineStartTimeFieldDoc),
+	}
+
+	w.tomb.Go(func() error {
+		defer close(w.outCh)
+		return w.loop()
+	})
+	return w
+}
+
+// Changes returns the event channel for the watcher.
+func (w *modelMachineStartTimeWatcher) Changes() <-chan []string {
+	return w.outCh
+}
+
+func (w *modelMachineStartTimeWatcher) loop() error {
+	docWatcher := newCollectionWatcher(w.backend, colWCfg{col: machinesC})
+	defer func() { _ = docWatcher.Stop() }()
+
+	var (
+		timer           = w.clk.NewTimer(w.quiesceInterval)
```

Comment:
I'm surprised that you create this with the timer started - you don't need to set the timer for the initial event since you send it immediately. If there aren't any events in the next `quiesceInterval` won't this fire unnecessarily?

Reply:
Not a huge deal obviously.

Reply:
I did this to avoid having to add this bit of code to stop the timer after creating it:

```go
if ! timer.Stop() {
  <- timer.Chan()
}
```

Worst case it will fire once and the `if len(unprocessedDocs) == 0` block will immediately return.